src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp

```diff
@@ -847,6 +851,16 @@ const WellTracerProperties& Well::getTracerProperties() const {
     return *this->tracer_properties;
 }
 
+const Well::WellProductivityIndex& Well::getWellProductivityIndex() const
+{
+    if (this->productivity_index)
+        return *this->productivity_index;
+    else
+        throw std::logic_error {
+            "WELPI not activated in well " + this->name()
```

Comment:
Couldn't this just as well return unity productivity index?

Reply:
> Couldn't this just as well return unity productivity index?

Alas, there's no sensible default here.  Since we're in SI at this point, `return 1.0` would essentially mean 1 SM3/(s * Pa)` which is a huge value even for gas.  The final CTF scaling value is the ratio of this value to a dynamically calculated one.  I suppose we could invent some kind of "no scaling" default value but that would also add a check to the caller.  I'm on the fence here.

Reply:
> > Couldn't this just as well return unity productivity index?
> 
> Alas, there's no sensible default here. Since we're in SI at this point, `return 1.0` would essentially mean 1 SM3/(s * Pa)` which is a huge value even for gas. The final CTF scaling value is the ratio of this value to a dynamically calculated one. I suppose we could invent some kind of "no scaling" default value but that would also add a check to the caller. I'm on the fence here.

Having thought about this some more, I believe I need to restructure the way we apply `WELPI` slightly.  I am going to split our current [`Well::applyWellProdIndexScaling()`](https://github.com/OPM/opm-common/blob/d0416d963d673b4e582bc7172a31cb19d68aee61/src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp#L814-L828) into two parts; one which calculates the scaling factor based on a ratio of stored data and a dynamically PI value, and one which applies a user-controlled scaling factor to eligible connections.  I think we need the latter, or something very much like it, in order to preserve the scaling effects of multiple `WELPI` keywords which apply to the same well at separate points through the simulation run.

@joakim-hove : Would you like to merge this PR now and then I'll create a new PR for that split or would you like me to make the split as part of this PR too?

Reply:
I have no strong feelings - but it sound like the current work is not really complete, and then I suggest to amend this PR with the necessary changes.

Reply:
> it sound like the current work is not really complete, and then I suggest to amend this PR with the necessary changes.

Very good.  I'll make those changes and update. For now I'm marking this as "work in progress" to reflect the status.

Reply:
> > it sound like the current work is not really complete, and then I suggest to amend this PR with the necessary changes.
> 
> Very good. I'll make those changes and update. For now I'm marking this as "work in progress" to reflect the status.

Those other changes turned out to take a new direction and become extensive enough to warrant a separate PR/review process.  I'll create that PR later today.  For now I'm marking this as ready for final review as a smaller component.